src/main/java/world/bentobox/greenhouses/greenhouse/BiomeRecipe.java

```diff
@@ -236,22 +236,30 @@ public void addReqBlocks(Material blockMaterial, int blockQty) {
 
     /**
      * Check if block should be converted
-     * @param gh  - greenhouse
      * @param b - block to check
      */
-    public void convertBlock(Greenhouse gh, Block b) {
-        conversionBlocks.get(b.getType()).stream().filter(Objects::nonNull)
-        .filter(bc -> random.nextDouble() < bc.getProbability())
-        .forEach(bc -> {
-            // Check if the block is in the right area, up, down, n,s,e,w
-            if (ADJ_BLOCKS.stream().map(b::getRelative)
-                    .filter(r -> gh.contains(r.getLocation()))
-                    .map(Block::getType)
-                    .anyMatch(m -> bc.getLocalMaterial() == null || m == bc.getLocalMaterial())) {
-                // Convert!
-                b.setType(bc.getNewMaterial());
+    public void convertBlock(Block b) {
+        Material bType  = b.getType();
+        // Check if there is a block conversion for this block, as while the rest of the method wont do anything if .get() returns nothing anyway it still seems to be quite expensive
+        if(conversionBlocks.keySet().contains(bType)) {
+            for(GreenhouseBlockConversions conversion_option : conversionBlocks.get(bType)) {
+
+                // Roll the dice before bothering with checking the surrounding block as I think it's more common for greenhouses to be filled with convertable blocks and thus this dice roll wont be "wasted"
+                if(ThreadLocalRandom.current().nextInt(10000) < conversion_option.getProbability()) {
```

Comment:
Am I missing something here? Won't `ThreadLocalRandom.current().nextInt(10000)` return an integer between 0 and 10000, but `conversion_option.getProbability()` returns a double?

Reply:
Yeah it would, forgot I left that in.